test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java

```diff
@@ -0,0 +1,134 @@
+                                             " doInBackground is finished");
+                }
+                System.out.println("Done");
+                doneLatch.countDown();
+            }
+        };
+
+        worker.addPropertyChangeListener(
+            new PropertyChangeListener() {
+                public void propertyChange(PropertyChangeEvent evt) {
+                    // Before doInBackground method is invoked,
+                    // SwingWorker notifies PropertyChangeListeners about the
+                    // property change to StateValue.STARTED
+                    if (doInBackgroundFinished.get() &&
+                         worker.getState() == SwingWorker.StateValue.STARTED) {
+                        throw new RuntimeException(
+                               "PropertyChangeListeners called with " +
+                               "state STARTED after doInBackground is invoked");
+                    }
+                    // Ensure the STARTED state is notified
+                    // before doInBackground started running
+                    if (doInBackgroundStarted.get() &&
+                         worker.getState() == SwingWorker.StateValue.STARTED) {
+                        throw new RuntimeException(
+                               "PropertyChangeListeners called with " +
+                               "state STARTED before doInBackground is finished");
+                    }
+                    // After the doInBackground method is finished
+                    // SwingWorker notifies PropertyChangeListeners
+                    // property change to StateValue.DONE
+                    if (doInBackgroundFinished.get() &&
+                            worker.getState() != SwingWorker.StateValue.DONE) {
+                        throw new RuntimeException(
+                            "PropertyChangeListeners called after " +
+                            " doInBackground is finished but before State changed to DONE");
+                    }
+                }
+            });
+        worker.execute();
+        Thread.sleep(WAIT_TIME * 3);
+
+        final long start = System.currentTimeMillis();
+        worker.cancel(true);
+        final long end = System.currentTimeMillis();
+
+        if ((end - start) > 100) {
+            throw new RuntimeException("Cancel took too long: "
+                                       + ((end - start) / 1000.0d) + " s");
+        }
+        if (!doneLatch.await(CLEANUP_TIME + 1000L, TimeUnit.MILLISECONDS)) {
+            throw new RuntimeException("done didn't complete in time");
+        }
+        System.out.println("doInBackground " + doInBackgroundFinished.get() +
+                           " getState " + worker.getState());
+        if (doInBackgroundFinished.get() &&
+              worker.getState() != SwingWorker.StateValue.DONE) {
+            throw new RuntimeException("doInBackground is finished " +
+                                       " but State is not DONE");
```

Comment:
```suggestion
                                       "but the State is not DONE");
```
Two consecutive spaces in an error message are redundant, one is enough.

The article seems missing?

Reply:
Modified except condition kept separate for distinct failure message